### PR TITLE
feat: Auto detect the use of Podman from DOCKER_HOST

### LINKED ISSUE #604: [Bug]: GenericContainer fails when used with podman, unless SkipReaper is true
### Testcontainers version

0.15.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host Arch

x86_64

### Go Version

1.19

### Docker version

```shell
mero@dibbler ~/tmp/x$ podman version
Version:      3.4.7
API Version:  3.4.7
Go Version:   go1.16.15
Built:        Thu Apr 21 19:38:09 2022
OS/Arch:      linux/amd64
```

### Docker info

```shell
mero@dibbler ~/tmp/x$ podman info   
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers:
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.32-1.fc34.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.32, commit: '
  cpus: 12
  distribution:
    distribution: fedora
    variant: workstation
    version: "34"
  eventLogger: journald
  hostname: dibbler
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.17.12-100.fc34.x86_64
  linkmode: dynamic
  logDriver: k8s-file
  memFree: 2583973888
  memTotal: 16543584256
  ociRuntime:
    name: crun
    package: crun-1.4.4-1.fc34.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.4.4
      commit: 6521fcc5806f20f6187eb933f9f45130c86da230
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc34.x86_64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 16936591360
  swapTotal: 16936591360
  uptime: 220h 47m 37.24s (Approximately 9.17 days)
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /home/mero/.config/containers/storage.conf
  containerStore:
    number: 8
    paused: 0
    running: 0
    stopped: 8
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/mero/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 4
  runRoot: /run/user/1000/containers
  volumePath: /home/mero/.local/share/containers/storage/volumes
version:
  APIVersion: 3.4.7
  Built: 1650562689
  BuiltTime: Thu Apr 21 19:38:09 2022
  GitCommit: ""
  GoVersion: go1.16.15
  OsArch: linux/amd64
  Version: 3.4.7
```

### What happened?

Using the sample below works when running in Docker, but fails if I use podman via `DOCKER_HOST="unix:///run/user/1000/podman/podman.sock"` (as [recommended here](https://golang.testcontainers.org/system_requirements/using_podman/)). See the attached log for an error message.

If I add `SkipReaper: true` to the request, it works. But that prints a large warning about what a bad idea that is.

### Relevant log output

```shell
2022/11/02 13:58:54 github.com/testcontainers/testcontainers-go - Connected to docker:  
  Server Version: 3.4.7
  API Version: 1.40
  Operating System: fedora
  Total Memory: 15777 MB
2022/11/02 13:58:54 Starting container id: 3eb8ac1a90f5 image: docker.io/testcontainers/ryuk:0.3.4
2022/11/02 13:58:54 Waiting for container id 3eb8ac1a90f5 image: docker.io/testcontainers/ryuk:0.3.4
2022/11/02 13:58:54 Error response from daemon: can only create exec sessions on running containers: container state improper, host port waiting failed: could not start container: creating reaper failed: failed to create container
exit status 1
```

### Additional Information

```go
package main

import (
	"context"
	"log"

	"github.com/testcontainers/testcontainers-go"
)

func main() {
	ctx := context.Background()

	c, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
		ContainerRequest: testcontainers.ContainerRequest{
			Image: "busybox:latest",
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	c.Terminate(ctx)
}
```

### PR DESCRIPTION
**Pull request by jtwatson**
## What does this PR do?

When running `testcontainers-go` with **Podman**, it seems that `ProviderType: ProviderPodman,` must **always** be specified.

The documentation [here](https://github.com/testcontainers/testcontainers-go/blame/main/docs/system_requirements/using_podman.md#L9) seems to contradict my experience.

Since **Podman 4** was released with a new network stack, it appears that the `default` network for **Podman** uses the `bridge` driver and is named `podman`.  As such, the library must be configured diffrently when using **Podman**.

This PR attempts to auto-detect when `ProviderType` has not been explicitly set and **Podman** is in use, using the following logic:

If `ProviderType` has not been set, and `DOCKER_HOST` contains `podman.sock`, then use `ProviderPodman`

## Why is it important?

When attempting to use `testcontainers-go` with **Podman**, it appears that it is always the case that 
```
	ProviderType: testcontainers.ProviderPodman,
```
must be set. When it is not set, the resulting error is quite difficult to troubleshoot. 

```
testcontainers.GenericContainer(): Error response from daemon: container create: unable to find network with name or ID bridge: network not found: creating reaper failed: failed to create container
```

If the library can reliably detect when Podman is in use, the developer experience would be much better.

## Related issues

- Closes #604

### CODE REVIEW

**Comment by mdelapenya** [comment]
@jtwatson at first sight, LGTM! Thanks for adding this automatic detection.

Could you update the docs for Podman describing what this PR does in a high level language? 🙏

File: docs/system_requirements/using_podman.md [hunk 1]
```diff
@@ -36,4 +35,47 @@ func TestSomething(t *testing.T) {
...
+
+## Fedora
+
+`DOCKER_HOST` environment variable must be set
+
+```
+> export DOCKER_HOST=unix://$XDG_RUNTIME_DIR/podman/podman.sock
+```
+
+SELinux may require a custom policy be applied to allow the reeper container to connect to and write to a socket. Once you experience the se-linux error, you can run the following commands to create and install a custom policy.
```
Comment (mdelapenya) [inline docs/system_requirements/using_podman.md:55]:
```suggestion
SELinux may require a custom policy be applied to allow the reaper container to connect to and write to a socket. Once you experience the se-linux error, you can run the following commands to create and install a custom policy.
```

**Comment by jtwatson** [comment]
@mdelapenya I question if the docs are accurate around when it is required to manually specify the `ProviderType`. I can't think of a situation where you would be using Docker and not need to define `DOCKER_HOST`.

It is probably good to leave it in so the usage is documented, but I wonder if the advice around when to use it should be corrected or not.

**Comment by mdelapenya** [comment]
> @mdelapenya I question if the docs are accurate around when it is required to manually specify the `ProviderType`. I can't think of a situation where you would be using Docker and not need to define `DOCKER_HOST`.
> 
> It is probably good to leave it in so the usage is documented, but I wonder if the advice around when to use it should be corrected or not.

I'm planning to move the logic of the provider, and many other things, to the testcontainer.properties, so it's defined when bootstrapping the library, not leaving the decision to manually add it to each container request.

**Comment by jtwatson** [comment]
If you would like I can add a test to validate the auto-detection based on the `DOCKER_HOST` value.

**Comment by jtwatson** [comment]
@mdelapenya I added unit tests, squashed my PR, and updated the title as a feature. Should be ready to go.

File: provider_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package testcontainers
+
+import (
+	"os"
+	"testing"
+)
+
+// TestProviderTypeGetProviderAutodetect should NOT be run in parallel as it sets environment variables.
```
Comment (mdelapenya) [inline provider_test.go:8]: Using `t.Setenv` would resolve that limitation

Comment (jtwatson) [reply provider_test.go:8]: Thanks! I didn't know about `t.Setenv()`. Nice to learn new tricks.

According to the docs, it still can't be used in Parallel tests, but it is a lot simpler. I will switch to using it.

File: docs/system_requirements/using_podman.md [hunk 3]
```diff
@@ -36,4 +35,47 @@ func TestSomething(t *testing.T) {
 }
 ```
 
-The `ProviderPodman` configures the `DockerProvider` with the correct default network for Podman to ensure also complex network scenarios are working as with Docker.
\ No newline at end of file
+The `ProviderPodman` configures the `DockerProvider` with the correct default network for Podman to ensure complex network scenarios are working as with Docker.
+
+## Podman socket activation
+
+The reaper container needs to connect to the docker daemon to reape containers, so the podman socket service must be started:
```
Comment (cristianrgreco) [inline docs/system_requirements/using_podman.md:42]:
```suggestion
The reaper container needs to connect to the docker daemon to reaper containers, so the podman socket service must be started:
```

Comment (jtwatson) [reply docs/system_requirements/using_podman.md:42]: It looks like I used an obsolete spelling of reap. But reaper is not correct in this context. I will fix it.

**Review by mdelapenya — APPROVED** [review]

**Comment by mdelapenya** [comment]
It's extremely weird that the neo4j module fails in a consistent manner on CI. I've fetched this branch locally, run the tests, and they pass 🤔

**Comment by mdelapenya** [comment]
> It's extremely weird that the neo4j module fails in a consistent manner on CI. I've fetched this branch locally, run the tests, and they pass 🤔

For some reason that inconsistency is no longer present 🤷

**Pull request merged**